src/server/pkg/storage/fileset/reader.go

```diff
@@ -53,20 +93,82 @@ func indexToContentHeader(idx *index.Header) {
 // Read reads from the current file in the tar stream.
 func (r *Reader) Read(data []byte) (int, error) {
 	// Lazily setup reader for underlying file.
-	if r.tr == nil {
-		if err := r.setupReader(); err != nil {
-			return 0, err
-		}
+	if err := r.setupReader(); err != nil {
+		return 0, err
 	}
 	return r.tr.Read(data)
 }
 
 func (r *Reader) setupReader() error {
-	r.cr.NextRange(r.hdr.Idx.DataOp.DataRefs)
-	r.tr = tar.NewReader(r.cr)
-	// Remove tar header from content stream.
-	if _, err := r.tr.Next(); err != nil {
-		return err
+	if r.tr == nil {
+		r.cr.NextRange(r.hdr.Idx.DataOp.DataRefs)
+		r.tr = tar.NewReader(r.cr)
+		// Remove tar header from content stream.
+		if _, err := r.tr.Next(); err != nil {
+			return err
+		}
 	}
 	return nil
 }
+
+// WriteToFiles does an efficient copy of files from a file set reader to writer.
+// The optional pathBound specifies the upper bound (exclusive) for files to copy.
+func (r *Reader) WriteToFiles(w *Writer, pathBound ...string) error {
+	for {
+		hdr, err := r.Peek()
+		if err != nil {
+			if err == io.EOF {
+				return nil
+			}
+			return err
+		}
+		if len(pathBound) > 0 && strings.Compare(hdr.Hdr.Name, pathBound[0]) >= 0 {
```

Comment:
Is there a reason you're using `strings.Compare` rather than just doing `hdr.Hdr.Name >= pathBound[0]`? I think that'd be easier to understand.